File: Project_Sourcecode_go/state/MDA_EFSM.go
package state

import (
	"cs586_project/output"
	"fmt"
)

type MdaEfsm struct {
	Attempts     int
	P            *output.OP
	StateQueue   []StateApi
	CurrentState StateApi
}

func (m *MdaEfsm) Init(p *output.OP) {
	m.P = p
	s := new(Start)
	s.M = m
	i := new(Idel)
	i.M = m
	c := new(CheckPin)
	c.M = m
	r := new(Ready)
	r.M = m
	s1 := new(S1)
	s1.M = m
	o := new(Overdrawn)
	o.M = m
	l := new(Locked)
	l.M = m
	su := new(Suspended)
	su.M = m
	m.StateQueue = []StateApi{
		s,
		i,
		c,
		r,
		s1,
		o,
		l,
		su,
	}

	m.CurrentState = m.StateQueue[0]
}

func (m *MdaEfsm) Open() {
	m.CurrentState.Open()
}

func (m *MdaEfsm) Login() {
	m.CurrentState.Login()
}

func (m *MdaEfsm) IncorrectLogin() {
	m.CurrentState.IncorrectLogin()
}

func (m *MdaEfsm) IncorrectPin(max int) {
	m.CurrentState.IncorrectPin(max)
}

func (m *MdaEfsm) CorrectPinBelowMin() {
	m.CurrentState.CorrectPinBelowMin()
}

func (m *MdaEfsm) CorrectPinAboveMin() {
	m.CurrentState.CorrectPinAboveMin()
}

func (m *MdaEfsm) Deposit() {
	m.CurrentState.Deposit()
}

func (m *MdaEfsm) AboveMinBalance() {
	m.CurrentState.AboveMinBalance()
}

func (m *MdaEfsm) BelowMinBalance() {
	m.CurrentState.BelowMinBalance()
}

func (m *MdaEfsm) Logout() {
	m.CurrentState.Logout()
}

func (m *MdaEfsm) Balance() {
	m.CurrentState.Balance()
}

func (m *MdaEfsm) Withdraw() {
	m.CurrentState.Withdraw()
}

func (m *MdaEfsm) WithdrawBelowMinBalance() {
	m.CurrentState.WithdrawBelowMinBalance()
}

func (m *MdaEfsm) Lock() {
	m.CurrentState.Lock()
}

func (m *MdaEfsm) IncorrectLock() {
	m.CurrentState.IncorrectLock()
}

func (m *MdaEfsm) Unlock() {
	m.CurrentState.Unlock()
}

func (m *MdaEfsm) IncorrectUnlock() {
	m.CurrentState.IncorrectUnlock()
}

func (m *MdaEfsm) Suspend() {
	m.CurrentState.Suspend()
}

func (m *MdaEfsm) Activate() {
	m.CurrentState.Activate()
}

func (m *MdaEfsm) Close() {
	m.CurrentState.Close()
}

func (m *MdaEfsm) NoFunds() {
	m.CurrentState.NoFunds()
}

func (m *MdaEfsm) ChangeState(index int) {
	m.CurrentState = m.StateQueue[index]
	fmt.Println("change state :" + m.CurrentState.GetName())
}

File: Project_Sourcecode_java/src/main/java/edu/iit/cs586/project/MDA_EFSM.java
package edu.iit.cs586.project;

import edu.iit.cs586.project.output.OP;
import edu.iit.cs586.project.state.*;

import java.util.ArrayList;
import java.util.List;

/**
 * @author CharmingLee <charminglee@jieztech.com>
 * @date 2022/5/1 20:23
 */
public class MDA_EFSM {
    private List<State> stateList;
    private State currentState;
    public int attempts = 0;
    public OP p;

    public MDA_EFSM (OP p) {
        this.p = p;
        stateList = new ArrayList<>();

        stateList.add(new Start(this));
        stateList.add(new Idel(this));
        stateList.add(new CheckPin(this));
        stateList.add(new Ready(this));
        stateList.add(new S1(this));
        stateList.add(new Overdrawn(this));
        stateList.add(new Locked(this));
        stateList.add(new Suspended(this));
        stateList.add(new Close(this));

        //init state
        changeState(0);
    }

    public void Open() {
        currentState.open();
    }

    public void Login() {
        currentState.login();
    }

    public void IncorrectLogin() {
        currentState.incorrectLogin();
    }

    public void IncorrectPin(int max) {
        currentState.incorrectPin(max);
    }

    public void CorrectPinBelowMin() {
        currentState.correctPinBelowMin();
    }

    public void CorrectPinAboveMin() {
        currentState.correctPinAboveMin();
    }

    public void Deposit() {
        currentState.deposit();
    }

    public void BelowMinBalance() {
        currentState.belowMinBalance();
    }

    public void AboveMinBalance() {
        currentState.aboveMinBalance();
    }

    public void Logout() {
        currentState.logout();
    }

    public void Balance() {
        currentState.balance();
    }

    public void Withdraw() {
        currentState.withdraw();
    }

    public void WithdrawBelowMinBalance() {
        currentState.withdrawBelowMinBalance();
    }

    public void NoFunds() {
        currentState.noFunds();
    }

    public void Lock() {
        currentState.lock();
    }

    public void IncorrectLock() {
        currentState.incorrectLock();
    }

    public void Unlock() {
        currentState.unlock();
    }

    public void IncorrectUnlock() {
        currentState.incorrectUnlock();
    }

    public void Suspend() {
        currentState.suspend();
    }

    public void Activate() {
        currentState.activate();
    }

    public void Close() {
        currentState.close();
    }

    public void changeState(int index) {
        this.currentState = stateList.get(index);
        System.out.println("MDA_EFSM changeState " + this.currentState.getName());
    }
}
